Add handlers for Byron migration

paweljakubas opened the pull request:
# Issue Number

#778 (`migrateByronWallet`)
#779 (`getByronWalletMigrationInfo`)

# Overview

This PR:

- [x] Adds a handler for `getByronWalletMigrationInfo`.
- [x] Adds a handler for `migrateByronWallet`.

# Not included in this PR (reserved for future PRs)

- More user-friendly error handling for the case where a user attempts to migrate _**from**_ a known Shelley wallet, or _**to**_ a known Byron wallet. (We can only migrate from Byron to Shelley.)
- A complete set of integration tests. We can add these in a further PR.


<!-- Additional comments or screenshots to attach if any -->

<!-- 
Don't forget to:

 ✓ Self-review your changes to make sure nothing unexpected slipped through
 ✓ Assign yourself to the PR
 ✓ Assign one or several reviewer(s)
 ✓ Once created, link this PR to its corresponding ticket
 ✓ Acknowledge any changes required to the Wiki
-->

paweljakubas committed:
Add incomplete handler with wallet helping function

KtorZ commented on lib/core/src/Cardano/Wallet.hs:
```diff
@@ -751,6 +754,34 @@ estimateTxFee ctx wid recipients = do
+
+-- | Estimate a Byron wallet migration cost by estimating fee by doing migration
+-- coin selection first
+estimateByronWalletMigrationCost
```
We shouldn't be talking about `Byron` here, this module is completely agnostic to different kind of wallet. So, let's simply call this one `estimateWalletMigrationCost` :+1:

paweljakubas replied:
done

paweljakubas committed:
finish impl of handler and rely on ErrNoSuchWallet

paweljakubas committed:
define ErrMigrationWallet

paweljakubas committed:
migrateByronWallet almost done

adopt most of review comments

KtorZ commented on lib/core/src/Cardano/Wallet/Api/Server.hs:
```diff
@@ -813,22 +817,71 @@ getByronWallet ctx wid =
+        liftHandler $ withWorkerCtx rndCtx sWid liftE3 $ \wrk ->
+            W.submitTx wrk sWid (tx, meta, wit)
+
+        now <- liftIO getCurrentTime
```
This could / should happen only once outside of the `forM` :+1:

paweljakubas replied:
corrected

KtorZ commented on lib/core/src/Cardano/Wallet/Api/Server.hs:
```diff
@@ -813,22 +817,71 @@ getByronWallet ctx wid =
+
+    -- (c) get coin selections
+    (_,coinSels) <- liftHandler $ withWorkerCtx rndCtx sWid throwE $ \wrk ->
+        W.computeWalletMigrationCostCoinSels wrk sWid
```
What about `W.createMigration` :p ?

paweljakubas replied:
changed the name :-)

KtorZ commented on lib/core/src/Cardano/Wallet.hs:
```diff
@@ -751,6 +756,32 @@ estimateTxFee ctx wid recipients = do
+    let fOpts = feeOpts tl (bp ^. #getFeePolicy)
+    let batchSize = fromIntegral $ idealBatchSize cOpts
+    let coinSels = selectCoinsForMigration fOpts batchSize utxo
+    let fees = fmap (estimateFee fOpts) coinSels
```
Although this computation should lead to the same result, I'd rather compute fee as the sum of `inputBalance sel - changeBalance sel` for all coin selections so we really reflect the actual fee!

paweljakubas replied:
corrected

KtorZ commented on lib/core/src/Cardano/Wallet.hs:
```diff
@@ -1065,6 +1096,18 @@ data ErrStartTimeLaterThanEndTime = ErrStartTimeLaterThanEndTime
+data ErrMigrationWallet
+    = ErrMigrationWalletNoSuchWallet ErrNoSuchWallet
+    | ErrMigrationWalletSeqWalletInSource
+    | ErrMigrationWalletRndWalletInDestination
```
These would probably need to have wallet ids as parameter to make for nicer error, no :thinking: ?

paweljakubas replied:
agree, added `WalletId`

KtorZ commented on lib/core/src/Cardano/Wallet/Api/Server.hs:
```diff
@@ -813,22 +817,71 @@ getByronWallet ctx wid =
+    let (ApiMigrateByronWalletData (ApiT pwd)) = migrateData
+    forM coinSels $ \selection -> do
+        (tx, meta, wit) <- liftHandler $ withWorkerCtx rndCtx sWid liftE2 $ \wrk ->
+            W.signTx wrk sWid pwd selection
```
I actually didn't even realize it'd play out so nicely with our already built-in functions :o ..

KtorZ commented on lib/core/src/Cardano/Wallet.hs:
```diff
@@ -751,6 +758,32 @@ estimateTxFee ctx wid recipients = do
+    let batchSize = fromIntegral $ idealBatchSize cOpts
+    let coinSels = selectCoinsForMigration fOpts batchSize utxo
+    let fees = fmap (\sel -> inputBalance sel - changeBalance sel) coinSels
+    pure (foldl (+) 0 fees, coinSels)
```
:+1:

KtorZ committed:
generate change addresses within the 'createMigration' function!

KtorZ committed:
fix migrateWallet function accordingly

KtorZ committed:
split back migration creation from execution (but still generate addresses for the target wallet!)

jonathanknowles committed:
Rename `createMigration` to `createMigrationSourceData`.

Additionally, clarify the comment.

This function doesn't actually initiate a migration, but it does create
the data necessary to perform a migration from a source wallet. Hence
adding the suffix "SourceData" makes sense here.

jonathanknowles committed:
Reformat according to coding standards.

https://github.com/input-output-hk/cardano-wallet/wiki/Coding-Standards#limit-line-length-to-80-characters

jonathanknowles committed:
Inline definitions of `liftE1` and `liftE2`.

These definitions are used only once, and the names don't add any
meaning.

jonathanknowles committed:
Rename `executeMigration` to `assignMigrationTargetAddresses`.

Additionally, clarify the comment.

This function doesn't actually execute a migration, but it does generate
and assign target addresses to an existing set of migration data.

jonathanknowles committed:
Move definition of `pwd` to `where` clause.

Its value only depends on the function arguments, and doesn't depend on
the monadic context.

jonathanknowles committed:
Acquire both random and sequential contexts before generating migration data.

jonathanknowles committed:
Provide implementation of `getByronWalletMigrationInfo` operation.

jonathanknowles committed:
Extract out function `selectionFee`.

jonathanknowles committed:
Make existing integration tests pass for `getByronWalletMigrationInfo` operation.

jonathanknowles committed:
Improve comments for `getByronWalletMigrationInfo` and `migrateByronWallet`.

jonathanknowles committed:
Move definition of `passphrase` to where clause.

It depends only on the function arguments, and not on the monadic
context.

jonathanknowles committed:
Use the constant `a` from `FeePolicy` as the value of `dustThreshold`.

KtorZ committed:
Remove need for having an encryption passphrase to generate change for SeqState
